File: Topics/01_Encapsulation/01_01_Evolution/cpp_source/evolution.cpp
#include <iostream>
#include <string.h>
using namespace std;

struct Account {
private:
    char name[12];
    char address[20];
    int phone;
    double balance;

public:
    // void chargeAccountFee(Account * this, double fee)
    void chargeFee(double fee) {
        this->balance = this->balance - fee;
        balance = balance - fee;
    }

    void withdraw(double amount) {
    }

    void deposit(double amount) {
    }

};

void chargeAccountFee(Account * acct, double fee) {
    acct->balance = acct->balance - fee;
}

char * getAccountName(Account * acct) {
    return acct->name;
}

void withdrawFromAccount(Account * acct, double amount) {

}

void depositIntoAccount(Account * acct, double amount) {

}

// void setName(Account * acct, char * newName)

int main(int argc, char ** argv) {

    Account jan;
    Account bob;

    jan.withdraw(10.0);
    jan.chargeFee(1.0);
    bob.deposit(10.0);
    bob.chargeFee(1.0);

    strcpy(jan.name, "Jan");
    jan.balance = 50.0;

    chargeAccountFee(&jan, 1.0);
    char * p = getAccountName(&bob);

    withdrawFromAccount(&jan, 10.0);
    chargeAccountFee(&jan, 1.0);

    depositIntoAccount(&bob, 10.0);
    chargeAccountFee(&bob, 1.0);

    cout << p << endl;

    cout << jan.balance;

}

File: Topics/01_Encapsulation/01_06_Memory_and_pointers/python_source/evolution.py
def do_fun():
    a = 5
    while a>0 and a!=10:   
        if a==3:
            continue        
        a = a - 1     
        
class Account:
    
    def _helper_function(self):
        pass
    
    def deposit(self, amount):
        self.balance = self.balance + amount
        
    def withdraw(self):
        pass
    
    def chargeFee(self,amount):
        pass
    
def say_hi_EN():
    print('Hello')
    
def say_hi_SP():
    print('Hola')
    
greet = say_hi_EN
greet()

greet = say_hi_SP
greet()

   
        
import dis
print( dis.dis(do_fun))